### PR TITLE
ESQL: Dependency check for binary plans

### PR DESCRIPTION
**Pull request by alex-spies**
The dependency checker for query plans should take into account binary plans and make sure that fields required from the left hand side are actually obtained from there (and analogously for the right).

### CODE REVIEW

File: docs/reference/esql/functions/description/categorize.asciidoc
Comment (alex-spies) [inline docs/reference/esql/functions/description/categorize.asciidoc:1]: Unrelated, but the regenerated docs for this were not committed, yet.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/PhysicalVerifier.java [hunk 1]
```diff
@@ -36,11 +35,6 @@ public Collection<Failure> verify(PhysicalPlan plan) {
         Failures depFailures = new Failures();
 
         plan.forEachDown(p -> {
-            if (p instanceof AggregateExec agg) {
-                var exclude = Expressions.references(agg.ordinalAttributes());
-                DEPENDENCY_CHECK.checkPlan(p, exclude, depFailures);
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/PhysicalVerifier.java:41]: Simplified by excluding the ordinal attributes already in `AggregateExec.references()`, which should be more correct, as these attributes are not required for the `AggregateExec` to be executed.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/AggregateExec.java [hunk 2]
```diff
@@ -184,7 +184,9 @@ public List<Attribute> output() {
 
     @Override
     protected AttributeSet computeReferences() {
-        return mode.isInputPartial() ? new AttributeSet(intermediateAttributes) : Aggregate.computeReferences(aggregates, groupings);
+        return mode.isInputPartial()
+            ? new AttributeSet(intermediateAttributes)
+            : Aggregate.computeReferences(aggregates, groupings).subtract(new AttributeSet(ordinalAttributes()));
```
Comment (bpintea) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/physical/AggregateExec.java:189]: 👍

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/PlanConsistencyChecker.java [hunk 3]
```diff
@@ -26,17 +28,26 @@ public class PlanConsistencyChecker<P extends QueryPlan<P>> {
...
+        } else if (p instanceof BinaryExec binaryExec) {
+            checkMissing(p, binaryExec.leftReferences(), binaryExec.left().outputSet(), "missing references from left hand side", failures);
+            checkMissing(
+                p,
+                binaryExec.rightReferences(),
+                binaryExec.right().outputSet(),
+                "missing references from right hand side",
+                failures
+            );
+        } else {
```
Comment (bpintea) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/PlanConsistencyChecker.java:49]: Nit: it'd be nice if we could deduplicate a bit here, maybe with a `checkBinary(P, AttributeSet, AttributeSet, String, Failures)`? But can stay as is too.

**Review by bpintea — APPROVED** [review]
Nice!

**Pull request merged**